### PR TITLE
EES-3399 UI test to Amend then cancel an amendment, check if the content is intact

### PR DESCRIPTION
**Pull request by N-moh**
UI test to check if all the content remains intact after making changes to release summary, replace data, edit data block, edit footnotes etc, then cancel amendment.
Screenshot (74)

### CODE REVIEW

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+
+    user clicks button    Choose an infographic as alternative
+    user chooses file    id:chartConfigurationForm-file    ${FILES_DIR}test-infographic.png
+    user checks radio is checked    Use table title
+    user enters text into element    id:chartConfigurationForm-alt    Sample alt text
+
+    user clicks button    Save chart options
+
+    user waits until page contains    Chart preview
+    user checks infographic chart contains alt    id:chartBuilderPreview    Sample alt text
```
Comment (duncan-at-hiveit) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:110]: Given that we're not changing the chart config as part of creating the amendment, I'd probably cut this part of the test out, as I don't think that it's adding much to the test case. I could be wrong! smile

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+
+    ${section}=    user gets accordion section content element    Dates test subject
+    user clicks link    Replace data    ${section}
+
+    user waits until h2 is visible    Data file details
+    user checks headed table body row contains    Subject title    Dates test subject
+    user checks headed table body row contains    Data file    dates.csv
+    user checks headed table body row contains    Metadata file    dates.meta.csv
+    user checks headed table body row contains    Number of rows    118    wait=%{WAIT_SMALL}
+    user checks headed table body row contains    Data file size    17 Kb    wait=%{WAIT_SMALL}
```
Comment (duncan-at-hiveit) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:227]: I would cut these verification lines out

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+    user checks headed table body row cell contains    Number of rows    1    118    wait=%{WAIT_SMALL}
+    user checks headed table body row cell contains    Data file size    1    17 Kb    wait=%{WAIT_SMALL}
+    user checks headed table body row cell contains    Status    1    Data replacement in progress    wait=%{WAIT_LONG}
+
+    user checks headed table body row cell contains    Subject title    2    Dates test subject
+    user checks headed table body row cell contains    Data file    2    dates-replacement.csv
+    user checks headed table body row cell contains    Metadata file    2    dates-replacement.meta.csv
+    user checks headed table body row cell contains    Number of rows    2    118    wait=%{WAIT_SMALL}
+    user checks headed table body row cell contains    Data file size    2    17 Kb    wait=%{WAIT_SMALL}
+    user checks headed table body row cell contains    Status    2    Complete    wait=%{WAIT_LONG}
```
Comment (duncan-at-hiveit) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:252]: I'd cut these out too

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+    user checks category checkbox is checked    Date    24/03/2020
+
+    user clicks element    id:filtersForm-submit
+    user waits until results table appears    %{WAIT_LONG}
+
+    user checks table column heading contains    1    1    2020 Week 13
+    user checks headed table body row cell contains    Number of open settings    1    23,000
+    user checks headed table body row cell contains    Proportion of settings open    1    2%
+    user checks headed table body row cell contains    Number of open settings    1    23,600
+    user checks headed table body row cell contains    Proportion of settings open    1    1%
```
Comment (duncan-at-hiveit) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:313]: I'd cut these out too

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+    user checks headed table body row contains    Status    Complete    wait=%{WAIT_LONG}
+
+Verify pre existing footnotes
+    user clicks link    Footnotes
+    user waits until h2 is visible    Footnotes
+    user waits until page contains element    testid:Footnote - Applies to all data 1
+    user waits until page contains element    testid:Footnote - Applies to all data 2
+    user waits until page does not contain element    testid:Footnote - Applies to all data 3
+
+Navigate to the 'Content' page to check if its intact
```
Comment (duncan-at-hiveit) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:405]: I'd maybe rename this to:

`Verify that the Release content is unchanged`

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+    user opens accordion section    Dates test subject
+    ${section}=    user gets accordion section content element    Dates test subject
+    user checks headed table body row contains    Subject title    Dates test subject
+    user checks headed table body row contains    Data file    dates.csv
+    user checks headed table body row contains    Metadata file    dates.meta.csv
+    user checks headed table body row contains    Number of rows    118    wait=%{WAIT_SMALL}
+    user checks headed table body row contains    Data file size    17 Kb    wait=%{WAIT_SMALL}
+    user checks headed table body row contains    Status    Complete    wait=%{WAIT_LONG}
+
+Verify pre existing footnotes
```
Comment (duncan-at-hiveit) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:398]: I'd maybe rename this to:

`Verify that the footnotes are unchanged`

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+
+    user checks list has x items    testid:footnotes    2    ${data_block_table}
+    user checks list item contains    testid:footnotes    1
+    ...    Applies to all data 1
+    ...    ${data_block_table}
+    user checks list item contains    testid:footnotes    2
+    ...    Applies to all data 2
+    ...    ${data_block_table}
+
+Verify Test text accordion section contains correct text as before
```
Comment (duncan-at-hiveit) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:445]: I'd maybe rename this to:

`Verify that the Test text accordion is unchanged`

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+    user waits until h2 is visible    Public data guidance
+
+    user waits until page contains element    id:dataGuidanceForm-content
+    user waits until page contains element    id:dataGuidance-dataFiles
+    user enters text into element    id:dataGuidanceForm-content    Test data guidance content
+    user waits until page contains accordion section    Dates test subject
+
+    user checks summary list contains    Filename    dates.csv
+    user checks summary list contains    Geographic levels    National
+    user checks summary list contains    Time period    2020 Week 13 to 2021 Week 24
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:44]: (as the same use case is covered in another test already)

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+    user clicks button    Upload file
+
+    user waits until page contains accordion section    Test ancillary file 1
+    user opens accordion section    Test ancillary file 1    id:file-uploads
+
+    ${section_1}=    user gets accordion section content element    Test ancillary file 1    id:file-uploads
+    user checks summary list contains    Title    Test ancillary file 1    ${section_1}
+    user checks summary list contains    Summary    Test ancillary file 1 summary    ${section_1}
+    user checks summary list contains    File    test-file-1.txt    ${section_1}
+    user checks summary list contains    File size    12 B    ${section_1}
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:66]: (as the same use case is covered in another test already)

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot (hunk 2 above)
Comment (duncan-at-hiveit) [reply tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:227]: (as the same use case is covered in another test already)

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot (hunk 3 above)
Comment (duncan-at-hiveit) [reply tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_cancel.robot:252]: (as the same use case is covered in another test already)

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+    user clicks link by visible text    Table    ${section}
+    user waits until parent contains element    ${section}
+    ...    xpath:.//*[@data-testid="dataTableCaption" and text()="Dates table title"]
+    user waits until parent contains element    ${section}    xpath:.//*[.="Source: Dates source"]
+    user checks table column heading contains    1    1    2020 Week 13    ${section}
+    user checks headed table body row cell contains    Number of open settings    1    22,900    ${section}
+    user checks headed table body row cell contains    Proportion of settings open    1    1%    ${section}
+    user closes accordion section    Dates data block    id:releaseMainContent
+
+Verify that the Dates data block table has footnotes are unchanged
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot:395]:
```suggestion
Verify that the Dates data block table footnotes are unchanged
```

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+    ...    Applies to all data 2
+    ...    ${data_block_table}
+
+Verify that the Test text accordion is unchanged
+    user opens accordion section    Test text    id:releaseMainContent
+    ${section}=    user gets accordion section content element    Test text    id:releaseMainContent
+    user waits until parent contains element    ${section}    xpath:.//p[text()="Some test text !"]
+    user closes accordion section    Test text    id:releaseMainContent
+
+Verify that the Embedded URL accordian section is unchanged
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot:413]:
```suggestion
Verify that the Embedded URL accordion section is unchanged
```

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+    user checks headed table body row contains    Status    Complete    wait=%{WAIT_LONG}
+
+Verify that the footnotes are unchanged
+    user clicks link    Footnotes
+    user waits until h2 is visible    Footnotes
+    user waits until page contains element    testid:Footnote - Applies to all data 1
+    user waits until page contains element    testid:Footnote - Applies to all data 2
+    user waits until page does not contain element    testid:Footnote - Applies to all data 3
+
+Verify that the Release content is unchanged
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot:367]:
```suggestion
Verify that release content is unchanged
```

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+    user clicks button    Save footnote
+    user waits until h2 is visible    Footnotes
+
+Confirm amendment has footnotes
+    user waits until h2 is visible    Footnotes
+    user waits until page contains element    testid:Footnote - Applies to all data 1
+    user waits until page contains element    testid:Footnote - Applies to all data 2
+    user waits until page contains element    testid:Footnote - Applies to all data 3
+
+User navigates to Data blocks page
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot:251]:
```suggestion
Navigate to Data blocks page
```

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+
+Navigate to newly published release page
+    user clicks link    ${PUBLICATION_NAME}
+    user waits until h1 is visible    ${PUBLICATION_NAME}    %{WAIT_MEDIUM}
+
+Verify release URL and page caption
+    user checks url contains    %{PUBLIC_URL}/find-statistics/ui-tests-publish-release-%{RUN_IDENTIFIER}
+    user waits until page contains title caption    ${RELEASE_NAME}
+
+Return to Admin and Create amendment
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot:182]:
```suggestion
Return to Admin and create amendment
```

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot [hunk 15]
```diff
@@ -0,0 +1,421 @@
...
+    ${publish_date_year}=    get current datetime    %Y
+    set suite variable    ${EXPECTED_PUBLISHED_DATE}
+    ...    ${publish_date_day} ${publish_date_month_word} ${publish_date_year}
+
+Verify newly published release is on Find Statistics page
+    user checks publication is on find statistics page    ${PUBLICATION_NAME}
+
+Navigate to newly published release page
+    user clicks link    ${PUBLICATION_NAME}
+    user waits until h1 is visible    ${PUBLICATION_NAME}    %{WAIT_MEDIUM}
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot:176]:
```suggestion
    user waits until h1 is visible    ${PUBLICATION_NAME}    %{WAIT_SMALL}
```

File: tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot [hunk 16]
```diff
@@ -0,0 +1,421 @@
...
+
+
+*** Variables ***
+${PUBLICATION_NAME}=    UI tests - publish release %{RUN_IDENTIFIER}
+${RELEASE_NAME}=        Financial Year 3000-01
+${DATABLOCK_NAME}=      Dates data block name
+
+
+*** Test Cases ***
+Create new publication for "UI tests topic" topic
```
Comment (luke-h1) [inline tests/robot-tests/tests/admin_and_public_2/bau/publish_amend_and_cancel.robot:22]:
```suggestion
Create new publication and release via API
```

**Review by luke-h1 — CHANGES_REQUESTED** [review]

**Comment by luke-h1** [comment]
Not too keen on the convention of using `Changes in response to PR comments` as a commit message. It doesn't really explain what PR comments or changes we've made. I'd suggest either rebasing to one commit or leaving a message such as `EES-xxxx - renamed variables and refactored x` (dependening on what you've changed of course)

**Review by luke-h1 — APPROVED** [review]

**Pull request closed without merging**